### PR TITLE
Adapter logging interface

### PR DESCRIPTION
**Pull request by nathaniel-may**
waiting for #4137 to merge first

### Description

Adapter maintainers currently use our old logging system. This exposes the new eventing system to them so that they have to make minimal changes to stay up-to-date for v1. Adapter maintainers will not have to touch every log call site to use the new system.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

**Comment by nathaniel-may** [comment]
I should probably add the `exception` log level here too.

**Comment by nathaniel-may** [comment]
re-requesting prior approvals because I made a significant change since then.

File: core/dbt/events/adapter_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        event.exc_info = exc_info
+        event.stack_info = stack_info
+        event.extra = extra
+
+        fire_event(event)
+
+    def exception(
+        self,
+        msg: str,
+        exc_info: Any = True,  # this default is what makes this method different
```
Comment (emmyoop) [inline core/dbt/events/adapter_endpoint.py:76]: I like this call out.

**Review by emmyoop — APPROVED** [review]

**Comment by nathaniel-may** [comment]
Should this object have some kind of doc string? What should it say?

**Review by leahwicz — APPROVED** [review]

File: core/dbt/events/README.md [hunk 2]
```diff
@@ -7,3 +7,16 @@ The event module provides types that represent what is happening in dbt in `even
...
+To integrate existing log messages from adapters, you likely have a line of code like this in your adapter already:
+```python
+from dbt.logger import GLOBAL_LOGGER as logger
+```
+
+Simply change it to these two lines with your adapter's database name, and all your existing call sites will now use the new system for v1.0:
+```python
+from dbt.events import AdapterLogger
+logger = AdapterLogger("<database name>")
+# e.g. AdapterLogger("Snowflake")
```
Comment (jtcohen6) [inline core/dbt/events/README.md:21]: 👍

**Review by jtcohen6 — APPROVED** [review]
The interface and explainer look great to me.

Could this PR also do the two-line work of switching over the Postgres adapter? It looks like there's one spot where it imports `dbt.logger` and should instead use `AdapterLogger("Postgres")`? That could be good as far as having a concrete example to point to when cutting over more adapters.

If you'd rather than come in a subsequent PR, that's also fine by me.

**Comment by nathaniel-may** [comment]
Logbook is causing issues when I moved the Postgres adapter over to the new interface. Instead of spinning my wheels investigating, I migrated all these changes into #4176 because Logbook isn't a problem over there.

**Comment by nathaniel-may** [comment]
Tests passed over there so I'm verifying that it works over here for record.

**Pull request closed without merging**